### PR TITLE
[SIMD] Refactor extend helper

### PR DESCRIPTION
**Pull request by MaxGraey**
Idea and motivation was described [here](https://github.com/WebAssembly/binaryen/pull/4016#issuecomment-885602111)

### CODE REVIEW

**Comment by MaxGraey** [comment]
For **I64x2 extensions in #4016 just needs add afterwards:
```cpp
Literal Literal::extendLowSToVecI64x2() const {
  return extend<2, int32_t, int64_t, LaneOrder::Low>(*this);
}
Literal Literal::extendHighSToVecI64x2() const {
  return extend<2, int32_t, int64_t, LaneOrder::Low>(*this);
}
Literal Literal::extendLowUToVecI64x2() const {
  return extend<2, uint32_t, uint64_t, LaneOrder::Low>(*this);
}
Literal Literal::extendHighUToVecI64x2() const {
  return extend<2, uint32_t, uint64_t, LaneOrder::Low>(*this);
}
```

@tlively wdyt?

File: src/wasm/literal.cpp [hunk 1]
```diff
@@ -2352,41 +2352,41 @@ Literal Literal::narrowUToVecI16x8(const Literal& other) const {
 
 enum class LaneOrder { Low, High };
 
-template<size_t Lanes,
-         LaneArray<Lanes * 2> (Literal::*IntoLanes)() const,
-         LaneOrder Side>
+template<size_t Lanes, typename LaneFrom, typename LaneTo, LaneOrder Side>
 Literal extend(const Literal& vec) {
-  LaneArray<Lanes* 2> lanes = (vec.*IntoLanes)();
+  LaneArray<Lanes* 2> lanes = getLanes<LaneFrom, Lanes * 2>(vec);
```
Comment (tlively) [inline src/wasm/literal.cpp:2357]:
```suggestion
  LaneArray<Lanes * 2> lanes = getLanes<LaneFrom, Lanes * 2>(vec);
```

File: src/wasm/literal.cpp [hunk 2]
```diff
@@ -2352,41 +2352,41 @@ Literal Literal::narrowUToVecI16x8(const Literal& other) const {
...
   LaneArray<Lanes> result;
   for (size_t i = 0; i < Lanes; ++i) {
-    result[i] = lanes[(Side == LaneOrder::Low) ? i : i + Lanes];
+    result[i] =
+      Literal((LaneTo)(LaneFrom)lanes[(Side == LaneOrder::Low) ? i : i + Lanes]
+                .geti32());
```
Comment (tlively) [inline src/wasm/literal.cpp:2362]: It might be worth have one or two helper variables in here just to break it up and make it more readable.

**Review by tlively — APPROVED** [review]
Looks great!

File: src/wasm/literal.cpp (hunk 1 above)
Comment (MaxGraey) [reply src/wasm/literal.cpp:2357]: Unfortunately formatter wants only `Lanes* 2` on the left side and only `Lanes * 2` on the right side 🤷

Comment (tlively) [reply src/wasm/literal.cpp:2357]: Ugh it probably thinks it's a pointer. Oh well.

**Comment by MaxGraey** [comment]
Done!

**Pull request merged**